File: mmupnp/src/main/java/net/mm2d/upnp/internal/manager/DeviceHolder.java
/*
 * Copyright (c) 2019 大前良介 (OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.upnp.internal.manager;

import net.mm2d.upnp.Device;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.TimeUnit;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;

/**
 * ControlPointで発見したDeviceを保持するクラス。
 *
 * <p>Deviceの有効期限を確認し、有効期限が切れたDeviceをLostとして通知する。
 *
 * @author <a href="mailto:ryo@mm2d.net">大前良介 (OHMAE Ryosuke)</a>
 */
public class DeviceHolder implements Runnable {
    private static final long MARGIN_TIME = TimeUnit.SECONDS.toMillis(10);

    public interface ExpireListener {
        void onExpire(@Nonnull Device device);
    }

    @Nonnull
    private final Object mThreadLock = new Object();
    private volatile boolean mShutdownRequest = false;
    @Nullable
    private Thread mThread;

    @Nonnull
    private final ExpireListener mExpireListener;
    @Nonnull
    private final Map<String, Device> mDeviceMap;

    /**
     * インスタンス作成。
     *
     * @param listener 期限切れの通知を受け取るリスナー
     */
    public DeviceHolder(@Nonnull final ExpireListener listener) {
        mDeviceMap = new LinkedHashMap<>();
        mExpireListener = listener;
    }

    /**
     * スレッドを開始する。
     */
    public void start() {
        mShutdownRequest = false;
        synchronized (mThreadLock) {
            mThread = new Thread(this, getClass().getSimpleName());
            mThread.start();
        }
    }

    /**
     * スレッドに割り込みをかけ終了させる。
     */
    public void shutdownRequest() {
        mShutdownRequest = true;
        synchronized (mThreadLock) {
            if (mThread != null) {
                mThread.interrupt();
                mThread = null;
            }
        }
    }

    /**
     * Device追加。
     *
     * @param device 追加されるDevice
     */
    public synchronized void add(@Nonnull final Device device) {
        mDeviceMap.put(device.getUdn(), device);
        notifyAll();
    }

    @Nullable
    public synchronized Device get(@Nonnull final String udn) {
        return mDeviceMap.get(udn);
    }

    /**
     * Device削除。
     *
     * @param device 削除されるDevice。
     * @return 削除されたDevice
     */
    public synchronized Device remove(@Nonnull final Device device) {
        return mDeviceMap.remove(device.getUdn());
    }

    /**
     * Device削除。
     *
     * @param udn 削除されるDeviceのudn。
     * @return 削除されたDevice
     */
    public synchronized Device remove(@Nonnull final String udn) {
        return mDeviceMap.remove(udn);
    }

    /**
     * 登録されたDeviceをクリア。
     */
    public synchronized void clear() {
        mDeviceMap.clear();
    }

    /**
     * 現在保持しているDeviceの順序を保持したリストを作成して返す。
     *
     * @return Deviceのリスト
     */
    @Nonnull
    public synchronized List<Device> getDeviceList() {
        return new ArrayList<>(mDeviceMap.values());
    }

    /**
     * 保持しているDeviceの数を返す。
     *
     * @return Deviceの数
     */
    public synchronized int size() {
        return mDeviceMap.size();
    }

    @Override
    public synchronized void run() {
        try {
            while (!mShutdownRequest) {
                while (mDeviceMap.size() == 0) {
                    wait();
                }
                expireDevice();
                waitNextExpireTime();
            }
        } catch (final InterruptedException ignored) {
        }
    }

    private void expireDevice() {
        final long now = System.currentTimeMillis();
        final Iterator<Device> i = mDeviceMap.values().iterator();
        while (i.hasNext()) {
            final Device device = i.next();
            if (device.getExpireTime() < now) {
                i.remove();
                mExpireListener.onExpire(device);
            }
        }
    }

    private void waitNextExpireTime() throws InterruptedException {
        if (mDeviceMap.size() == 0) {
            return;
        }
        final long sleep = findMostRecentExpireTime() - System.currentTimeMillis() + MARGIN_TIME;
        wait(Math.max(sleep, MARGIN_TIME)); // 負の値となる可能性を排除
    }

    private long findMostRecentExpireTime() {
        long recent = Long.MAX_VALUE;
        for (final Device device : mDeviceMap.values()) {
            final long expireTime = device.getExpireTime();
            if (recent > expireTime) {
                recent = expireTime;
            }
        }
        return recent;
    }
}

File: mmupnp/src/main/java/net/mm2d/upnp/internal/manager/DeviceHolder.kt
/*
 * Copyright (c) 2019 大前良介 (OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.upnp.internal.manager

import net.mm2d.upnp.Device
import net.mm2d.upnp.internal.thread.TaskExecutors
import java.util.concurrent.FutureTask
import java.util.concurrent.TimeUnit
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock
import kotlin.math.max

/**
 * A class holding [Device] found in [net.mm2d.upnp.ControlPoint].
 *
 * Check the expiration date of Device, and notify the expired Device as Lost.
 *
 * @author [大前良介 (OHMAE Ryosuke)](mailto:ryo@mm2d.net)
 *
 * @constructor initialize
 * @param expireListener Listener to receive expired notifications
 */
internal class DeviceHolder(
    private val taskExecutors: TaskExecutors,
    private val expireListener: (Device) -> Unit
) : Runnable {
    private var futureTask: FutureTask<*>? = null
    private val threadLock = ReentrantLock()
    private val lock = ReentrantLock()
    private val condition = lock.newCondition()
    private val deviceMap = mutableMapOf<String, Device>()

    val deviceList: List<Device>
        get() = lock.withLock {
            return deviceMap.values.toList()
        }
    val size: Int
        get() = lock.withLock {
            return deviceMap.size
        }

    fun start() {
        threadLock.withLock {
            FutureTask(this, null).also {
                futureTask = it
                taskExecutors.manager(it)
            }
        }
    }

    fun stop() {
        threadLock.withLock {
            futureTask?.cancel(false)
            futureTask = null
        }
    }

    private fun isCanceled(): Boolean {
        return futureTask?.isCancelled ?: true
    }

    fun add(device: Device) {
        lock.withLock {
            deviceMap[device.udn] = device
            condition.signalAll()
        }
    }

    operator fun get(udn: String): Device? {
        lock.withLock {
            return deviceMap[udn]
        }
    }

    fun remove(device: Device): Device? {
        lock.withLock {
            return deviceMap.remove(device.udn)
        }
    }

    fun remove(udn: String): Device? {
        lock.withLock {
            return deviceMap.remove(udn)
        }
    }

    fun clear() {
        lock.withLock {
            deviceMap.clear()
        }
    }

    override fun run() {
        Thread.currentThread().let {
            it.name = it.name + "-device-holder"
        }
        lock.withLock {
            try {
                while (!isCanceled()) {
                    while (deviceMap.isEmpty()) {
                        condition.await()
                    }
                    expireDevice()
                    waitNextExpireTime()
                }
            } catch (ignored: InterruptedException) {
            }
        }
    }

    private fun expireDevice() {
        val now = System.currentTimeMillis()
        deviceMap.values
            .toList()
            .filter { it.expireTime < now }
            .forEach {
                deviceMap.remove(it.udn)
                expireListener.invoke(it)
            }
    }

    @Throws(InterruptedException::class)
    private fun waitNextExpireTime() {
        if (deviceMap.isEmpty()) {
            return
        }
        val sleep = max(
            findMostRecentExpireTime() - System.currentTimeMillis() + MARGIN_TIME,
            MARGIN_TIME
        ) // avoid negative value
        condition.await(sleep, TimeUnit.MILLISECONDS)
    }

    private fun findMostRecentExpireTime(): Long {
        return deviceMap.values.minBy { it.expireTime }
            ?.expireTime ?: 0L
    }

    companion object {
        private val MARGIN_TIME = TimeUnit.SECONDS.toMillis(10)
    }
}
